### DISCUSSION TITLE
Customize the perspective

### DISCUSSION THREAD

**Discussion by idoric** [cetz-package/cetz#746] [Q&A] [answered]
By default in CeTZ the perspective is such that¹: the z axis forms an angle of 45° with the horizontal, and the lengths along this axis are multiplied by the coefficient very close to sqrt(2)/2 (or at least I have every reason to suppose so). This is a rather good choice by default because this is how we do it on our student notebooks, by following the squares. But this is not what I want, I need two variants: 1) the angle of 45° and a coefficient of 0.5, and 2) an angle of 30° and a coefficient of 0.7 (but sqrt(2)/2 could do the trick I suppose). For the first option, I found a solution with `scale(z: 0.707106781)` (because L × (sqrt(2)/2) × (sqrt(2)/2) = L × 1/2). For the second option I'm stuck, I looked at transformation matrices, but I don't even understand why it's a 4×4 matrix and not 3×3 (and even if I did, I've never done this kind of thing). Any ideas?

¹ Sorry for not using more precise vocabulary, but from the little I have read the vocabulary in this area does not seem to have the same meaning in French and English (starting with the cavalier perspective), so I prefer not to use it.

**Comment by johannes-wolf** [comment]
Does the `ortho` env. do what you want? (https://cetz-package.github.io/docs/api/draw-functions/projections/ortho).

**Reply by idoric** [reply]
I may have misunderstood how the ortho() function works, but I don't think so. With `x: 10deg, y: 0deg` the x and y axes are in the plane of the screen, but as soon as I change the y parameter to try to get the z axis where I want it, the x axis (and that makes sense) is no longer in the plane of the screen, and touching the z parameter only led to confusion for me 😬.

**Reply by johannes-wolf** [reply]
Can you give an example (image) of the perspective you want to achieve?
I thought you only want to change the `x:` rotation and keep the defaults of the other axes?

**Reply by idoric** [reply]
On the attachment, a cube with a side of 2 cm represented three times, from left to right: with the default perspective of CeTZ, with what is called in France “perspective cavalière avec un angle de fuite 45° et un rapport de réduction 0,5”, and with what is called in France “perspective cavalière avec un angle de fuite 30° et un rapport de réduction 0,7”.

![cube different projections](https://github.com/user-attachments/assets/e4cfb704-3d3c-4906-9f42-bae09f2a375e)

**Reply by johannes-wolf** [reply, marked as answer]
Ah ok, you may want something like this: 
```typst
#import "@preview/cetz:0.3.1"
#cetz.canvas({
  import cetz.draw: *

  grid((-1,-1), (12,4))

  let cube(pos) = scope({
    stroke(red + 2pt)
    set-origin(pos)

    on-xz({
      rect((0,0), (2,2))
      rect((0,0,-2), (2,2,-2))
    })
    
    rect((0,0), (2,2))
    rect((0,0,2), (2,2,2))
  })

  let persp(a, c: calc.sqrt(2)/2) = {
    set-transform(
      ((1, 0, calc.cos(a) * c,0),
       (0,-1,-calc.sin(a) * c,0),
       (0, 0, 0,  0),
       (0, 0, 0,  1)))
  }

  persp(45deg)
  cube((0,0))

  persp(45deg, c: .5)
  cube((4,0))
  
  persp(30deg)
  cube((8,0))
})
```

**Reply by johannes-wolf** [reply]
![grafik](https://github.com/user-attachments/assets/4a34f817-1dc3-4fc4-9463-4d0b3c4e3f05)

**Reply by idoric** [reply]
This is just perfect. Many thanks for this persp() function which will be very useful to me!

Small subsidiary question: am I wrong to assume that the last column of the matrix relates to the layer of the point?

**Reply by johannes-wolf** [reply]
No, it is not the layer. Here is an explanation of why 4x4 matrices are used for transformations: https://gamedev.stackexchange.com/questions/72044/why-do-we-use-4x4-matrices-to-transform-things-in-3d. The last column is needed for translation.

**Comment by uzgit** [comment]
Is there a way to apply this type of perspective on an image? I have added my image as content, but it is unaffected by the call to persp.

E.g.,
```
persp(45deg)
content((0,0), image("images/test.png"))
```

**Reply by johannes-wolf** [reply]
Typst hast `skew`, `rotate` and `scale` but does not support passing a 3×3 transformation matrix.
Would you be open to support a generic `transform` that accepts a 3x3 Matrix in Typst @laurmaedje?

Edit: 2x2 matrices.

**Reply by laurmaedje** [reply]
Not sure which exact kind of Matrix (PDF matrices have just six entries), but in principle yes.
